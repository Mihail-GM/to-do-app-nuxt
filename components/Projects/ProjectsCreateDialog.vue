<template>
  <v-dialog
    v-model="showDialog"
    max-width="600px"
    persistent
  >
    <v-card>
      <v-card-title class="primary mb-4">
        <span class="headline white--text">
          Create new project
        </span>

        <v-spacer />

        <v-btn
          id="app-dialog-hide-button"
          icon
          @click.prevent="closeDialog"
        >
          <v-icon color="white">
            mdi-close
          </v-icon>
        </v-btn>
      </v-card-title>

      <v-card-text>
        <v-container>
          <v-row>
            <v-col>
              <v-text-field
                id="title-text-field"
                v-model="project.name"
                :error-messages="nameErrors"
                data-q-name-input
                label="Project name*"
                required
                @blur="$v.project.name.$touch()"
              />
            </v-col>
          </v-row>
        </v-container>

        <small>*indicates required field</small>
      </v-card-text>

      <v-card-actions>
        <v-spacer />
        <v-btn
          id="close-project-dialog-button"
          class="white--text"
          color="red darken-1"
          data-q-close-button
          large
          @click.native="closeDialog"
        >
          Close
        </v-btn>

        <v-btn
          id="create-project-button"
          :disabled="$v.$anyError"
          color="primary darken-1"
          data-q-submit-button
          large
          @click.native="createProject"
        >
          Save
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script>
import { v4 as uuidv4 } from 'uuid'
import { maxLength, required } from 'vuelidate/lib/validators'
import { validationMixin } from 'vuelidate'

export default {
  name: 'ProjectsCreateDialog',
  mixins: [validationMixin],

  props: {
    showDialogProp: {
      require: true,
      default: false,
      type: Boolean
    }
  },

  validations: {
    project: {
      name: {
        required,
        maxLength: maxLength(32)
      }
    }
  },

  data () {
    return {
      project: {
        name: null,
        tasks: []
      }
    }
  },

  computed: {
    showDialog: {
      get () {
        return this.showDialogProp
      },
      set (value) {
        this.$emit('update:showDialogProp', value)
      }
    },

    nameErrors () {
      const errors = []
      if (!this.$v.project.name.$dirty) {
        return errors
      }

      !this.$v.project.name.maxLength && errors.push('Can contain maximum of 32 characters')
      !this.$v.project.name.required && errors.push('Must contain at least 1 character')
      return errors
    }
  },

  methods: {
    resetProject () {
      this.project = {
        name: null,
        tasks: []
      }
    },

    closeDialog () {
      this.resetProject()
      this.$emit('update:showDialogProp', false)
    },

    createProject () {
      this.$v.$touch()

      if (this.$v.$anyError) {
        return
      }

      this.$emit('createProject', { id: uuidv4(), ...this.project })
      this.closeDialog()
    }
  }
}

</script>
