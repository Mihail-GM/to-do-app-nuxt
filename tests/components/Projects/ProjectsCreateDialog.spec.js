import merge from 'lodash.merge'
import { v4 as uuidv4 } from 'uuid'
import flushPromises from 'flush-promises'
import ProjectsCreateDialog from '~/components/Projects/ProjectsCreateDialog.vue'

import {
  addVuelidate,
  addVuetify,
  addVuex,
  bootstrapVueContext,
  compositeConfiguration,
  customWrapper
} from '~/tests/utils'

jest.mock('uuid')

describe('ProjectsCreateDialog', () => {
  let vueContext = null

  beforeEach(() => {
    vueContext = bootstrapVueContext(
      compositeConfiguration(addVuelidate, addVuex, addVuetify)
    )
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  const storeMock = {
    state: {},
    dispatch: jest.fn(),
    commit: jest.fn()
  }

  const createWrapper = (overrides) => {
    const defaultMountingOptions = {
      mocks: {
        $store: storeMock
      },
      propsData: {
        showDialogProp: true
      }
    }

    return customWrapper(ProjectsCreateDialog, vueContext, merge(defaultMountingOptions, overrides)
    )
  }

  const findInputNameField = wrapper => wrapper.find('[data-q-name-input]')
  const findCloseButton = wrapper => wrapper.find('[data-q-close-button]')
  const findCreateButton = wrapper => wrapper.find('[data-q-submit-button]')

  describe('Input name text field', () => {
    test('Shows correct error message when name is not valid', async () => {
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('')
      await findInputNameField(wrapper).trigger('blur')
      await flushPromises()

      expect(wrapper.findAll('.v-messages__message').at(0).text()).toBe('Must contain at least 1 character')
    })

    test('Shows correct error message when name is more then 32 chars', async () => {
      expect.assertions(1)
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii')
      await findInputNameField(wrapper).trigger('blur')
      await flushPromises()

      expect(wrapper.findAll('.v-messages__message').at(0).text()).toBe('Can contain maximum of 32 characters')
    })

    test('Do not show error message when name is valid', async () => {
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('Valid project name')
      await findInputNameField(wrapper).trigger('blur')
      await flushPromises()

      expect(wrapper.findAll('.v-messages__message').length).toBe(0)
    })
  })

  describe('Close button', () => {
    test('Reset data and correct emit when close dialog', async () => {
      expect.assertions(3)
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('test')
      await findCloseButton(wrapper).trigger('click')
      await flushPromises()

      expect(wrapper.vm.project.name).toEqual(null)
      expect(wrapper.emitted('update:showDialogProp').length).toBe(1)
      expect(wrapper.emitted('update:showDialogProp').at(0)).toEqual([false])
    })
  })

  describe('Create project button', () => {
    test('Sets button to disabled when name is not valid', async () => {
      expect.assertions(1)
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('')
      await findInputNameField(wrapper).trigger('blur')
      await flushPromises()

      expect(findCreateButton(wrapper).attributes('disabled')).toBe('disabled')
    })

    test('Sets button to be clickable when name is valid', async () => {
      expect.assertions(1)
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('Project name')
      await findInputNameField(wrapper).trigger('blur')
      await flushPromises()

      expect(findCreateButton(wrapper).attributes('disabled')).toBe(undefined)
    })

    test('Calls correct emit when create is clicked with valid data', async () => {
      expect.assertions(2)
      uuidv4.mockImplementationOnce(() => '9c55f43d-49b7-49c2-bdf3-31706b2fb373')

      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('Project name')
      await findCreateButton(wrapper).trigger('click')
      await flushPromises()

      expect(wrapper.emitted('createProject').length).toBe(1)
      expect(wrapper.emitted('createProject').at(0)).toStrictEqual([
        {
          id: '9c55f43d-49b7-49c2-bdf3-31706b2fb373',
          name: 'Project name',
          tasks: []
        }
      ])
    })
  })

  describe('Snapshot tests', () => {
    test('Matches snapshot when component is mounted', () => {
      const wrapper = createWrapper()

      expect(wrapper.html()).toMatchSnapshot()
    })

    test('Matches snapshot when project name is incorrect', async () => {
      expect.assertions(1)
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('')
      await findInputNameField(wrapper).trigger('blur')

      expect(wrapper.html()).toMatchSnapshot()
    })

    test('Matches snapshot when project name is correct', async () => {
      expect.assertions(1)
      const wrapper = createWrapper()
      await findInputNameField(wrapper).setValue('Correct project name')
      await findInputNameField(wrapper).trigger('blur')

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
