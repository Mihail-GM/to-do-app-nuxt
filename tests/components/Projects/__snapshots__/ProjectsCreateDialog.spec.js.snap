// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectsCreateDialog Snapshot tests Matches snapshot when component is mounted 1`] = `
"<div class=\\"v-dialog__container\\">
  <div role=\\"dialog\\" aria-modal=\\"true\\" class=\\"v-dialog__content v-dialog__content--active\\" style=\\"z-index: 202;\\">
    <transition-stub name=\\"dialog-transition\\" appear=\\"true\\">
      <div tabindex=\\"0\\" class=\\"v-dialog v-dialog--active v-dialog--persistent\\" style=\\"transform-origin: center center; max-width: 600px;\\">
        <div class=\\"v-card v-sheet theme--light\\">
          <div class=\\"v-card__title primary mb-4\\"><span class=\\"headline white--text\\">
        Create new project
      </span>
            <div class=\\"spacer\\"></div> <button type=\\"button\\" class=\\"v-btn v-btn--icon v-btn--round theme--light v-size--default\\" id=\\"app-dialog-hide-button\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-close theme--light white--text\\"></i></span></button></div>
          <div class=\\"v-card__text\\">
            <div class=\\"container\\">
              <div class=\\"row\\">
                <div class=\\"col\\">
                  <div class=\\"v-input theme--light v-text-field\\">
                    <div class=\\"v-input__control\\">
                      <div class=\\"v-input__slot\\">
                        <div class=\\"v-text-field__slot\\"><label for=\\"title-text-field\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: absolute;\\">Project name*</label><input data-q-name-input=\\"\\" required=\\"required\\" id=\\"title-text-field\\" type=\\"text\\"></div>
                      </div>
                      <div class=\\"v-text-field__details\\">
                        <div class=\\"v-messages theme--light\\">
                          <transition-group-stub tag=\\"div\\" name=\\"message-transition\\" class=\\"v-messages__wrapper\\"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div> <small>*indicates required field</small>
          </div>
          <div class=\\"v-card__actions\\">
            <div class=\\"spacer\\"></div> <button type=\\"button\\" class=\\"white--text v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--large red darken-1\\" id=\\"close-project-dialog-button\\" data-q-close-button=\\"\\"><span class=\\"v-btn__content\\">
        Close
      </span></button> <button type=\\"button\\" class=\\"v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--large primary darken-1\\" id=\\"create-project-button\\" data-q-submit-button=\\"\\"><span class=\\"v-btn__content\\">
        Save
      </span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>"
`;

exports[`ProjectsCreateDialog Snapshot tests Matches snapshot when project name is correct 1`] = `
"<div class=\\"v-dialog__container\\">
  <div role=\\"dialog\\" aria-modal=\\"true\\" class=\\"v-dialog__content v-dialog__content--active\\" style=\\"z-index: 202;\\">
    <transition-stub name=\\"dialog-transition\\" appear=\\"true\\">
      <div tabindex=\\"0\\" class=\\"v-dialog v-dialog--active v-dialog--persistent\\" style=\\"transform-origin: center center; max-width: 600px;\\">
        <div class=\\"v-card v-sheet theme--light\\">
          <div class=\\"v-card__title primary mb-4\\"><span class=\\"headline white--text\\">
        Create new project
      </span>
            <div class=\\"spacer\\"></div> <button type=\\"button\\" class=\\"v-btn v-btn--icon v-btn--round theme--light v-size--default\\" id=\\"app-dialog-hide-button\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-close theme--light white--text\\"></i></span></button></div>
          <div class=\\"v-card__text\\">
            <div class=\\"container\\">
              <div class=\\"row\\">
                <div class=\\"col\\">
                  <div class=\\"v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field\\">
                    <div class=\\"v-input__control\\">
                      <div class=\\"v-input__slot\\">
                        <div class=\\"v-text-field__slot\\"><label for=\\"title-text-field\\" class=\\"v-label v-label--active theme--light\\" style=\\"left: 0px; position: absolute;\\">Project name*</label><input data-q-name-input=\\"\\" required=\\"required\\" id=\\"title-text-field\\" type=\\"text\\"></div>
                      </div>
                      <div class=\\"v-text-field__details\\">
                        <div class=\\"v-messages theme--light\\">
                          <transition-group-stub tag=\\"div\\" name=\\"message-transition\\" class=\\"v-messages__wrapper\\"></transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div> <small>*indicates required field</small>
          </div>
          <div class=\\"v-card__actions\\">
            <div class=\\"spacer\\"></div> <button type=\\"button\\" class=\\"white--text v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--large red darken-1\\" id=\\"close-project-dialog-button\\" data-q-close-button=\\"\\"><span class=\\"v-btn__content\\">
        Close
      </span></button> <button type=\\"button\\" class=\\"v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--large primary darken-1\\" id=\\"create-project-button\\" data-q-submit-button=\\"\\"><span class=\\"v-btn__content\\">
        Save
      </span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>"
`;

exports[`ProjectsCreateDialog Snapshot tests Matches snapshot when project name is incorrect 1`] = `
"<div class=\\"v-dialog__container\\">
  <div role=\\"dialog\\" aria-modal=\\"true\\" class=\\"v-dialog__content v-dialog__content--active\\" style=\\"z-index: 202;\\">
    <transition-stub name=\\"dialog-transition\\" appear=\\"true\\">
      <div tabindex=\\"0\\" class=\\"v-dialog v-dialog--active v-dialog--persistent\\" style=\\"transform-origin: center center; max-width: 600px;\\">
        <div class=\\"v-card v-sheet theme--light\\">
          <div class=\\"v-card__title primary mb-4\\"><span class=\\"headline white--text\\">
        Create new project
      </span>
            <div class=\\"spacer\\"></div> <button type=\\"button\\" class=\\"v-btn v-btn--icon v-btn--round theme--light v-size--default\\" id=\\"app-dialog-hide-button\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-close theme--light white--text\\"></i></span></button></div>
          <div class=\\"v-card__text\\">
            <div class=\\"container\\">
              <div class=\\"row\\">
                <div class=\\"col\\">
                  <div class=\\"v-input v-input--has-state theme--light v-text-field error--text\\">
                    <div class=\\"v-input__control\\">
                      <div class=\\"v-input__slot\\">
                        <div class=\\"v-text-field__slot\\"><label for=\\"title-text-field\\" class=\\"v-label theme--light error--text\\" style=\\"left: 0px; position: absolute;\\">Project name*</label><input data-q-name-input=\\"\\" required=\\"required\\" id=\\"title-text-field\\" type=\\"text\\"></div>
                      </div>
                      <div class=\\"v-text-field__details\\">
                        <div class=\\"v-messages theme--light error--text\\" role=\\"alert\\">
                          <transition-group-stub tag=\\"div\\" name=\\"message-transition\\" class=\\"v-messages__wrapper\\">
                            <div class=\\"v-messages__message\\">Must contain at least 1 character</div>
                          </transition-group-stub>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div> <small>*indicates required field</small>
          </div>
          <div class=\\"v-card__actions\\">
            <div class=\\"spacer\\"></div> <button type=\\"button\\" class=\\"white--text v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--large red darken-1\\" id=\\"close-project-dialog-button\\" data-q-close-button=\\"\\"><span class=\\"v-btn__content\\">
        Close
      </span></button> <button type=\\"button\\" class=\\"v-btn v-btn--disabled v-btn--has-bg theme--light v-size--large\\" id=\\"create-project-button\\" data-q-submit-button=\\"\\" disabled=\\"disabled\\"><span class=\\"v-btn__content\\">
        Save
      </span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>"
`;
